doc/LAYERS.org

```diff
@@ -0,0 +1,588 @@
+configuration you need to do should be doable within the =:init= or =:config=
+blocks of such a call.
+
+What goes where? Since =:init= is executed before load and =:config= after,
+these rules of thumb apply.
+
+In =:config= should be
+- Anything that requires the package to be already loaded.
+- Anything that takes a long time to run, which would ruin startup performance.
+
+The =:init= block should contain setup for the entry points to the package. This
+includes keybindings, if the package should be loaded manually by the user, or
+hooks, if the package should be loaded upon some event. It is not unusual to
+have both!
+
+** Use-package hooks
+Spacemacs includes a macro for adding more code to the =:init= or =:config=
+blocks of a call to =use-package=, after the fact. This is useful for =pre-init=
+or =post-init= functions to "inject" code into the =use-package= call of the
+=init= function.
+
+#+begin_src emacs-lisp
+  (spacemacs|use-package-add-hook helm
+    :pre-init
+    ;; Code
+    :post-init
+    ;; Code
+    :pre-config
+    ;; Code
+    :post-config
+    ;; Code
+    )
+#+end_src
+
+Since a call to =use-package= may evaluate the =:init= block immediately, any
+function that wants to inject code into this block must run =before= the call to
+=use-package=. Further, since this call to =use-package= typically takes place
+in the =init-<package>= function, calls to =spacemacs|use-package-add-hook=
+typically happen in the =pre-init-<package>= functions, and not in
+=post-init-<package>=.
+
+** Best practices
+If you break any of these rules, you should know what you are doing and have a
+good reason for doing it.
+
+*** Package ownership
+Each package should be owned by one layer only. The layer that owns the
+package should define its =init= function. Other layers should rely on
+=pre-init= or =post-init= functions.
+
+*** Localize your configuration
+*Each function can only assume the existence of one package.* With some
+exceptions, the =pre-init=, =init= and =post-init= functions can /only/
+configure exactly the package they are defined for. Since the user can exclude
+an arbitrary set of packages, there is no /a priori/ safe way to assume that
+another package is included. Use =configuration-layer/package-usedp= if you
+must.
+
+This can be very challenging, so please take this as a guideline and not
+something that is absolute. It is quite possible for the user to break her
+Spacemacs installation by excluding the wrong packages, and it is not our
+intention to prevent this at all costs.
+
+*** Don't assume a fixed order
+In Spacemacs, layers are loaded in order of inclusion in the dotfile, and
+packages are loaded in alphabetical order. You should not depend on the latter,
+and in most cases you should not depend on the former, either. Preferably, write
+your layer so that it is independent of load ordering. The =pre=- and
+=post-init= functions are helpful, together with =configuration-layer/package-usedp=.
+
```

Comment:
You are right about this since stuff can be lazy loaded but it is guaranteed that packages are configured alphabetically so when you know what you are doing you can rely on this property. It is the cases at several places in Spacemacs.

But maybe we should not document this to keep things simple. Relying on configuration order may be considered to be some kind of optimization we can do post merge of a PR.